File: test-gigachat-token.js
import axios from 'axios';
import https from 'https';
import fs from 'fs';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';
import dotenv from 'dotenv';

// Загружаем переменные окружения
dotenv.config({ path: '.env.local' });

// Настройка сертификатов
const certPath = path.resolve(process.cwd(), 'russian_trusted_root_ca.crt');
if (fs.existsSync(certPath)) {
  process.env.NODE_EXTRA_CA_CERTS = certPath;
  console.log('✅ Сертификат НУЦ Минцифры загружен:', certPath);
} else {
  console.log('⚠️ Сертификат НУЦ Минцифры не найден, используем отключение проверки SSL');
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
}

async function testGigaChatToken() {
  try {
    console.log('=== ТЕСТ ПОЛУЧЕНИЯ ТОКЕНА GIGACHAT ===');
    
    const authKey = process.env.GIGACHAT_API_KEY;
    const clientId = process.env.GIGACHAT_CLIENT_ID;
    
    console.log('Auth Key present:', !!authKey);
    console.log('Client ID present:', !!clientId);
    console.log('Auth Key (first 20 chars):', authKey ? authKey.substring(0, 20) + '...' : 'NOT FOUND');
    
    if (!authKey) {
      throw new Error('GIGACHAT_API_KEY не найден в переменных окружения');
    }
    
    const rqUID = uuidv4();
    console.log('Generated RqUID:', rqUID);
    
    console.log('Отправляем запрос к GigaChat OAuth...');
    console.log('URL: https://ngw.devices.sberbank.ru:9443/api/v2/oauth');
    
    const response = await axios.post('https://ngw.devices.sberbank.ru:9443/api/v2/oauth', 
      'scope=GIGACHAT_API_PERS',
      {
        headers: {
          'Authorization': `Basic ${authKey}`,
          'RqUID': rqUID,
          'Content-Type': 'application/x-www-form-urlencoded',
          'Accept': 'application/json'
        },
        httpsAgent: new https.Agent({
          rejectUnauthorized: false,
          secureProtocol: 'TLSv1_2_method',
          ca: fs.existsSync(certPath) ? fs.readFileSync(certPath) : undefined
        }),
        timeout: 30000, // 30 секунд таймаут
        maxRedirects: 5
      }
    );
    
    console.log('✅ УСПЕХ! Токен получен');
    console.log('Status:', response.status);
    console.log('Headers:', response.headers);
    console.log('Data:', response.data);
    
    if (response.data.access_token) {
      console.log('Access Token (first 50 chars):', response.data.access_token.substring(0, 50) + '...');
      console.log('Expires at:', new Date(response.data.expires_at).toISOString());
    }
    
    return response.data;
  } catch (error) {
    console.error('❌ ОШИБКА при получении токена:');
    console.error('Error type:', error.constructor.name);
    console.error('Error message:', error.message);
    
    if (error.response) {
      console.error('Response status:', error.response.status);
      console.error('Response headers:', error.response.headers);
      console.error('Response data:', error.response.data);
    } else if (error.request) {
      console.error('Request was made but no response received');
      console.error('Request details:', error.request);
    }
    
    console.error('Full error:', error);
    throw error;
  }
}

// Запускаем тест
testGigaChatToken()
  .then(result => {
    console.log('🎉 Тест завершен успешно!');
    process.exit(0);
  })
  .catch(error => {
    console.error('💥 Тест завершен с ошибкой!');
    process.exit(1);
  });

File: test_gigachat_token.py
import requests
import os
from dotenv import load_dotenv

# Загружаем переменные окружения
load_dotenv('.env.local')

def test_gigachat_token():
    url = "https://ngw.devices.sberbank.ru:9443/api/v2/oauth"
    
    # Получаем Authorization key из переменных окружения
    auth_key = os.getenv('GIGACHAT_API_KEY')
    
    if not auth_key:
        print("❌ GIGACHAT_API_KEY не найден в переменных окружения")
        return False
    
    print(f"✅ Authorization key найден: {auth_key[:20]}...")
    
    payload = {
        'scope': 'GIGACHAT_API_PERS'
    }
    
    headers = {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Accept': 'application/json',
        'RqUID': 'ac160d04-0557-4dc0-bc66-8e7875d69102',
        'Authorization': f'Basic {auth_key}'
    }
    
    print("Отправляем запрос к GigaChat OAuth...")
    print(f"URL: {url}")
    print(f"Headers: {headers}")
    
    try:
        response = requests.post(url, headers=headers, data=payload, timeout=30)
        
        print(f"✅ Статус ответа: {response.status_code}")
        print(f"Ответ: {response.text}")
        
        if response.status_code == 200:
            data = response.json()
            if 'access_token' in data:
                print(f"✅ Токен получен успешно!")
                print(f"Access Token (первые 50 символов): {data['access_token'][:50]}...")
                print(f"Expires at: {data.get('expires_at')}")
                return data['access_token']
            else:
                print("❌ Токен не найден в ответе")
                return False
        else:
            print(f"❌ Ошибка: {response.status_code}")
            return False
            
    except requests.exceptions.RequestException as e:
        print(f"❌ Ошибка запроса: {e}")
        return False

if __name__ == "__main__":
    print("=== ТЕСТ ПОЛУЧЕНИЯ ТОКЕНА GIGACHAT (Python) ===")
    token = test_gigachat_token()
    
    if token:
        print("🎉 Тест завершен успешно!")
    else:
        print("💥 Тест завершен с ошибкой!")

